### ISSUE TITLE
New random functions

### ISSUE THREAD

**Issue by agoose77** [scikit-hep/awkward#1230]
### Description of new feature

Hi all, 

I needed some Awkward-aware random functions today, and implemented some simple routines for `choice`, `shuffle`, and `random`. Whilst it is reasonably simple to implement these using `recursively_apply` and `broadcast_and_apply`, e.g.
```python3
def random(size=None, highlevel=True, behavior=None):
    if size is None:
        size = 1

    if isinstance(size, int):
        return ak.from_numpy(
            np.random.random(size=size), highlevel=highlevel, behavior=behavior
        )

    def getfunction(layout, depth):
        if not isinstance(layout, ak.layout.NumpyArray):
            return

        size = np.asarray(layout)
        offset = np.zeros(size.size + 1, dtype=np.uint64)
        offset[1:] = np.cumsum(size)

        return lambda: ak.layout.ListOffsetArray64(
            ak.layout.Index64(offset),
            ak.layout.NumpyArray(np.random.random(size=offset[-1])),
        )

    layout = ak.to_layout(size, allow_other=False)
    out = ak._util.recursively_apply(layout, getfunction)
    return ak._util.maybe_wrap_like(out, size, behavior, highlevel)
```    
these routines are important enough that I feel they deserve their own submodule in the Awkward namespace, e.g. `ak.random`. 
Equally, I don't think we want to implement _too_ many functions from NumPy. Every function we re-implement increases our maintenance burden. I believe that these three functions cover many of the use cases:
- sampling with replacement: `choice(array, size=None, axis=-1, rng=None)`
- sampling without replacement: `shuffle(array, axis=-1, rng=None)`
- random floats: `random(size=None, axis=-1, rng=None)`

In each of these cases, `size` is either an `int`, `None`, or an `ak.Array`.

We could also add `randint`, which is easily implemented in terms of `random`, so wouldn't be another kernel.

This issue is just a "to-do" tracker for my benefit for now. Does anyone have any thoughts?

**Comment by jpivarski** [comment]
This is a good idea: #489!

I think a nice way to do it would be to put the implementations in a submodule the same way that `ak.operations.*` (and `ak._v2.operations.*`) are, but then exposed the functions as a submodule, rather than merging names in a common namespace, the way that the others are. NumPy has established a tradition of randomization functions being in a submodule. (Many of these would be NEP-18 overloads of NumPy functions, anyway.)

**Comment by agoose77** [comment]
I agree that we should use a separate submodule, that would be an aesthetically familiar choice.

**Comment by agoose77** [comment]
I won't be working on this for a little bit, but I wanted to think about it up-front.

In NumPy, the preferred API is to use a `Generator`, whose methods the user can call to generate random values. Given that we currently support NumPy `>=1.13.1`, we would also need to implement the top-level legacy API. 

To implement the new generator API, I propose implementing the class `ak.random.Generator` which accepts an RNG from CuPy/NumPy,.  I think it will be necessary for the user to pass in the RNG instance from CuPy/NumPy: unlike most of the other Awkward highlevel functions, random generators in some cases are not required to accept any array-like arguments, e.g. `np.random.random(size=10)`, which means we cannot use `nplike` to determine the array implementation transparently.

nplike should gain a `default_rng()` method, that passes through to `np.default_rng()` or `cupy.default_rng()`. Maybe we will want to implement a `GeneratorLike` proxy that ensures we can only call a subset of methods, which would currently be:
- `random`
- `permutation`
- `randint`

**Comment by jpivarski** [comment]
This sounds good. Random number generators are different from other kinds of functions because they can have zero arguments (and yet return different values). Adding methods to nplike is fine—that's what it's there for. It shouldn't be too hard to make a `default_rng` for TypeTracer and JAX as well.

-----------------

What minimal version of NumPy could we instead take so that we don't have to maintain a legacy API?

The current minimum of 1.13.1 (Jul 6, 2017) was chosen because this is when [NEP 13](https://numpy.org/neps/nep-0013-ufunc-overrides.html) was introduced. Without that, innocent-looking expressions would return unexpected results, so it's a hard minimum.

Another meaningful point would be 1.17.0 (Jul 26, 2019), when [NEP 18](https://numpy.org/neps/nep-0018-array-function-protocol.html) was added. Currently, people can use Awkward Array with NumPy < 1.17.0, but they have to explicitly choose things like `ak.concatenate` instead of `np.concatenate`. If they use the latter, it would probably just be slow: I think NumPy would iterate over the ak.Array and create a corresponding np.ndarray, then use that. If there are any non-NumPy features, it would fail with an error message. I considered that to be not as catastrophic as returning wrong results.

If supporting randomization functions gets easier after some NumPy version number and it's not much newer than 3 years old, we can increase our minimum NumPy version.

(If there's ever reason to increase our minimum NumPy to 1.20 (Jan 30, 2021) or above, we'll need to drop Python 3.6 support. But requiring NumPy to only be 1 year old is too narrow of a window; it would make Awkward hard to install in some places.)

**Comment by agoose77** [comment]
It seems as though 1.16 was the final minor version for the legacy random API:
> The RandomState provides access to legacy generators. This generator is considered frozen and will have no further improvements. It is guaranteed to produce the same values as the final point release of NumPy v1.16.

1.17 still contains the legacy API, but importantly introduces this new API. There were some bugfixes in the 1.17 series, but I think the case for supporting 1.17.0 vs 1.17.x is that the end-user should determine whether they need those fixes.

I think Awkward could (and perhaps should) be opinionated enough to only implement the Generator interface.

**Comment by jpivarski** [comment]
A constraint of 1.17.0 versus 1.17.5 or something is probably not too different from a user's point of view. Typically, a system will have or have access to the latest minor release in a major sequence. If the new API wasn't "right" until 1.17.x for some x, then it would probably simplify people's lives by cutting on the right x.

So if we require NumPy >= 1.17.x, then implementing randomization functions will only need a Generator interface? If I understand correctly that setting this minimum NumPy means less work, then let's do it! I'm going to start a discussion on Deprecations about increasing this minimum.

**Comment by jpivarski** [comment]
#1238

**Comment by jpivarski** [comment]
Pinging @Duchstf here because I'm closing #489 in favor of this one. (Technically, this one is the duplicate because it came later, but I've already put it in a priority list.)
